File: C++/Demo1/Workspace/ApplicationFramework/src/ConsoleLog.cpp
#include "ConsoleLog.hpp"
#include <iostream>

#ifdef _WIN64
#include <windows.h>
#endif

using std::cout;
using std::endl;

namespace af
{
    ConsoleLog::ConsoleLog()
    {
#ifdef _WIN64
        consoleHandle = GetStdHandle(STD_OUTPUT_HANDLE);
        SetConsoleOutputCP(65001);
#else
        std::locale::global(std::locale("en_US.UTF-8"));
#endif
    }

    ConsoleLog::~ConsoleLog()
    {
        Reset();
    }

    void ConsoleLog::Reset()
    {
#ifdef _WIN64
        SetConsoleTextAttribute(consoleHandle, 7);
#else
        cout << "\033[0m";
#endif
    }

    void ConsoleLog::Print(const LogMessage& logMessage)
    {
        if (logMessage.LogLevel == LogMessage::LogLevel::Information) // Green
        {
#ifdef _WIN64
            SetConsoleTextAttribute(consoleHandle, 2);
#else
            cout << "\033[32m";
#endif
            cout << InformationLabel;
        }

        else if (logMessage.LogLevel == LogMessage::LogLevel::Warning) // Yellow
        {
#ifdef _WIN64
            SetConsoleTextAttribute(consoleHandle, 6);
#else
            cout << "\033[33m";
#endif
            cout << WarningLabel;
        }

        else if (logMessage.LogLevel == LogMessage::LogLevel::Error) // Red
        {
#ifdef _WIN64
            SetConsoleTextAttribute(consoleHandle, 4);
#else
            cout << "\033[31m";
#endif
            cout << ErrorLabel;
        }

        Reset();
        cout << logMessage.SourceName;

        if (!logMessage.Message.empty())
        {
            cout << sourceMessageSeparator << logMessage.Message;
        }

        cout << endl;
    }
}

File: C#/Demo2/Apps/IOBoard/IOBoardServer/ConsoleLog.cs
using System;



namespace IOBoardServer
{
    public class ConsoleLog : ILog
    {
        public event Action<string> ErrorOccurred;


        ConsoleColor defaultColor;


        public ConsoleLog()
        {
            defaultColor = Console.ForegroundColor;
        }

        public void NewMessage(LogLevels logLevel, string message)
        {
            switch(logLevel)
            {
                case LogLevels.Information:
                    Console.ForegroundColor = ConsoleColor.Green;
                    Console.Write("Information: ");
                    Console.ForegroundColor = defaultColor;
                    Console.WriteLine(message);
                    break;

                case LogLevels.Warning:
                    Console.ForegroundColor = ConsoleColor.Yellow;
                    Console.Write("Warning: ");
                    Console.ForegroundColor = defaultColor;
                    Console.WriteLine(message);
                    break;

                case LogLevels.Error:
                    Console.ForegroundColor = ConsoleColor.Red;
                    Console.Write("Error: ");
                    Console.ForegroundColor = defaultColor;
                    Console.WriteLine(message);
                    ErrorOccurred?.Invoke(message);
                    break;
            }
        }
    }
}
